lib/THC/THCThreadLocal.c

```diff
@@ -0,0 +1,43 @@
+#include "THCThreadLocal.h"
+#include "THCGeneral.h"
+
+
+THCThreadLocal THCThreadLocal_alloc()
+{
+#ifndef _WIN32
+  pthread_key_t key;
+  THAssert(pthread_key_create(&key, NULL) == 0);
+  return key;
+#else
+  DWORD key = TlsAlloc();
+  THAssert(key != TLS_OUT_OF_INDEXES);
+  return key;
+#endif
+}
+
+void THCThreadLocal_free(THCThreadLocal local)
+{
+#ifndef _WIN32
+  THAssert(pthread_key_delete(local) == 0);
+#else
+  THAssert(TlsFree(local));
+#endif
+}
+
+void* THCThreadLocal_get(THCThreadLocal local)
```

Comment:
This is an internal class, and we're using C++, so can we just use the templates to determine the held type. I understand that we might save on one or two malloc/free calls, but I'm not sure if it's worth complicating the code. Let's just store an `int*` pointing to a single value and return it a properly typed pointer/reference from get. If we really want to avoid the malloc, we can just specialize this class for small types (and hide all intptr_t casts inside this class).